torch/csrc/jit/codegen/cuda/test/test_gpu_fused_reduction.cpp

```diff
@@ -2170,6 +2311,618 @@ TEST_F(NVFuserTest, FusionCrossIterationGroupedGridAllreduce4_CUDA) {
+  tv4->computeAt(tv7, -1, ComputeAtMode::MostInlined);
+
+  // // This forces the grid reduction into the persistent loop which gives
+  // terrible perf tv3->computeAt(tv7, -1, ComputeAtMode::MostInlined);
+
+  // Inline residual path with output
+  tv5->computeAt(tv7, -1, ComputeAtMode::MostInlined);
+
+  fusion.printMath();
+  fusion.printKernel();
+
+  auto options_half = at::TensorOptions().dtype(at::kHalf).device(at::kCUDA, 0);
+  at::manual_seed(0);
+
+  const std::vector<int64_t> input_shape{N, H, W, C};
+  auto t0 = at::randn(input_shape, options_half);
+
+  FusionExecutor fe;
+  fe.compileFusion(&fusion, {t0});
+  auto cg_outputs = fe.runFusion({t0});
+  for (int i = 0; i < 10; ++i) {
+    clearL2Cache();
+    cg_outputs = fe.runFusion({t0});
+  }
+
+  auto t1 = t0.to(at::kFloat);
+  auto t2 = t1.sum({0, 1, 2});
+  auto t3 = t2.unsqueeze(0).unsqueeze(0).unsqueeze(0);
+  auto t4 = t1 - t3;
+
+  testValidate(&fusion, cg_outputs, {t0}, {t4}, __LINE__, __FILE__, "");
+}
+
+// Simple persistent kernel on outer normalization for schedule testing.
+TEST_F(NVFuserTest, FusionGridPersistOuterNormLike2_CUDA) {
+  std::vector<bool> bcast_pattern{true, true, true, false};
+  std::vector<int> reduction_dims{2, 1, 0};
+
+  std::unique_ptr<Fusion> fusion_ptr = std::make_unique<Fusion>();
+  Fusion& fusion = *fusion_ptr.get();
+  FusionGuard fg(&fusion);
+
+  auto tv0 = makeContigTensor(4, DataType::Half);
+  fusion.addInput(tv0);
+
+  // Manually project persistent buffer
+  auto tv1 = set(tv0);
+  // First use of persistent buffer, cast to float
+  auto tv2 = castOp(DataType::Float, tv1);
+  auto tv3 = sum(tv2, reduction_dims);
+  auto tv4 = broadcast(tv3, bcast_pattern);
+  // Second use of manually projected persistent buffer
+  auto tv5 = castOp(DataType::Float, tv1);
+  auto tv6 = sub(tv5, tv4);
+  auto tv7 = castOp(DataType::Half, tv6);
+  fusion.addOutput(tv7);
+
+  // [rN*H*W, C]
+  // Target
+  //      18 blocks per grid allreduce to
+  //      64 values of C at a time, with vector size of 4
+  //      16 threadIdx.x and 16 threadIdx.y
+  //      192 registes/thread -> 48 registers / thread for the reduction
+  //      dimension
+  //
+  // Target input size: 27648 x 4096
+
+  // Break down the outer norm 27648 x 4096 into 4 dimensions like channels last
+  // BN
+  int64_t N = 108, H = 16, W = 16, C = 4096;
+
+  int tidy = 16;
+  int tidx = 16;
+  int vect = 4;
+  int persist_buffer = 96;
+  int bidx = 4;
+
+  // [N, H, W, C]
+  // Reuse persistent buffer as the tv0 cache
+  auto tv0_cache = tv1;
+  auto tv7_cache = tv7->cacheBefore();
+
+  // [N, H, W, C]
+
+  tv3->merge(0)->merge(0);
+
+  tv3->split(0, tidy);
+  tv3->axis(1)->parallelize(ParallelType::TIDy);
+  // Persistent buffer
+  tv3->split(0, persist_buffer);
+  tv3->axis(0)->parallelize(ParallelType::BIDy);
+
+  // [rO{BIDy}, r56{persist}, rI{TIDy}, i]
+  // Vector dim
+  tv3->split(3, vect);
+  tv3->axis(4)->parallelize(ParallelType::Vectorize);
```

Comment:
This isn't needed, right?

Reply:
It's needed for the sort to get the iter domains in the right order.